python/pyarrow/dataset.py

```diff
@@ -437,7 +437,7 @@ def _filesystem_dataset(source, schema=None, filesystem=None,
     FileSystemDataset
     """
     format = _ensure_format(format or 'parquet')
-    partitioning = _ensure_partitioning(partitioning)
+    partitioning = _ensure_partitioning(partitioning, schema)
```

Comment:
This doesn't seem right to me, since the user-specified `schema` is meant to be the full dataset schema, while the `partitioning` functions takes the schema of the only the partitioning (although it is then a bit surprising that no tests fails). 

Is this change needed to fix the issue (apart from the fix in C++)? 
(on first sight, it doesn't seem needed to have the python test you added passing)

Reply:
Yes this wasn't needed for the Python test case I added in this PR. But it was needed for the previous one when the schema is specified. If it is not specified even with this change it works fine. But when specified I get `null` values in the `exp_meta` column.